### DISCUSSION TITLE
ピンの位置情報をrealmに保存したい

### DISCUSSION THREAD

**Discussion by mm0o21** [lit-iphone-course/lit-iphone-course#46] [Question-original] [answered]
(1) 実現したいこと
初期画面に表示されるピンのボタンを押したときにピンの位置がrealmに保存されるようにしたい
(2) 実現するために自分が試した内容とその結果
ボタンを押すと"unrecognized selector sent to instance"と表示されてしまい、アプリがクラッシュしてしまいます。関連付けを確認したところエラーが出ていたため関連づけし直したのですが、それでもまだ落ちてしまうため原因がわからない状態です。
(3) 調べた内容や資料
[https://qiita.com/ruwatana/items/7cfc63ce102938962260](https://github.com/lit-iphone-course/lit-iphone-course/discussions/url)
こちらの記事を参考に繋ぎ直してみたのですが治りませんでした
(4) 原因だと思われそうな箇所(ソースコードをコピペして貼る)

```swift
@IBAction func addPin(_ sender: UITapGestureRecognizer){
        
        if(sender.state != UIGestureRecognizer.State.began){
               return
           }

        //現在地
        let coordinate = mapView.userLocation.coordinate
            //ピンを生成
            let pin = MKPointAnnotation()
            //ピンに現在地を設定
            pin.coordinate = coordinate
            //mapにピンを表示
            mapView.addAnnotation(pin)
        
        coordinatesArray.append(["name":"ピン", "lat": mapView.userLocation.coordinate.latitude, "lon": mapView.userLocation.coordinate.longitude])
        makeMap()
        
        let location:CGPoint = sender.location(in: mapView)
        let center = mapView.convert(location, toCoordinateFrom: mapView)
        let lati:String = center.latitude.description
        let long:String = center.longitude.description
        
        savePin(latitude: lati, longitude: long)
    }
```
(5) GitHubのリンク
[https://github.com/mm0o21/sanphoto](https://github.com/lit-iphone-course/lit-iphone-course/discussions/url)

**Comment by txxxxc** [comment]
多分リポジトリがprivateになっているのが原因で見れないので、権限ほしいです🙏🙏

**Reply by mm0o21** [reply]
[https://github.com/mm0o21/sanphoto](url)

確認したところpublicになっていたのですが、私も質問の時のリンクからだと飛べませんでした、、
一旦URL貼り直してみます！

＜追記＞これもリンク先飛べないです

**Reply by yukiny0811** [reply]
mm0o21のユーザーページに飛べば見れた！

**Comment by yukiny0811** [comment, marked as answer]
Storyboard見た感じ、このピンのボタンを押したときに「segueでの画面遷移」と「addPin()」が同時に呼ばれちゃってるみたいだから、
<img width="488" alt="image" src="https://user-images.githubusercontent.com/28947703/192081802-1562865e-cd3a-4d62-9a7b-020a15dcad52.png">

下のgifみたいな感じで、Segueをボタンからドラッグするんじゃなくて、ViewControllerのFirstResponderからドラッグして次の画面に繋げ直そう！
![GIF 2022-09-24 at 2 37 14 PM](https://user-images.githubusercontent.com/28947703/192081949-32f32ad4-548b-498f-b8a6-e9acd611629d.gif)

そしたら、新しく繋げ直したsegueにidentifierをつけてあげて、
<img width="838" alt="image" src="https://user-images.githubusercontent.com/28947703/192081898-503c6dd7-98e9-4935-9639-bb1202bc1ac4.png">

addPin()の一番最後に画面遷移のコードをかいてあげよう！
<img width="750" alt="image" src="https://user-images.githubusercontent.com/28947703/192081926-de6cbcde-a7a3-47a4-923b-d2812541ad5e.png">

**Reply by mm0o21** [reply]
segue繋ぎ直してみましたがまだunrecognized selector sent to instanceと表示されて落ちてしまいます・・・！

**Reply by yukiny0811** [reply]
あ！ごめん！あとこのエラーの直接的な原因は、UIButtonのActionとしてaddPin()を宣言しているのに、senderとしてUITapGestureRecognizerを使っちゃってるところ！本当は```addPin(_ sender: UIButton)```にしなきゃいけない！
![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/28947703/192082418-9f155725-bcb9-4d14-b4f7-20bae7b90253.png)

もし、「ピンのボタンをタップしたときに、画面の中心の座標を使って次の画面に遷移する」っていう実装にしたかったら、
```.swift
@IBAction func addPin(_ sender: UIButton){
        let coordinate = mapView.userLocation.coordinate
        let pin = MKPointAnnotation()
        pin.coordinate = coordinate
        mapView.addAnnotation(pin)
        coordinatesArray.append(["name":"ピン", "lat": mapView.userLocation.coordinate.latitude, "lon": mapView.userLocation.coordinate.longitude])
        print(coordinatesArray)
        makeMap()
        
        //ここ大事
        let location:CGPoint = CGPoint(x: self.mapView.frame.width / 2, y: self.mapView.frame.height / 2)
        
        let center = mapView.convert(location, toCoordinateFrom: mapView)
        let lati:String = center.latitude.description
        let long:String = center.longitude.description
        savePin(latitude: lati, longitude: long)
        self.performSegue(withIdentifier: "next", sender: nil)
    }
```

こんな感じに変更しなきゃいけない！

**Reply by mm0o21** [reply]
なるほど！sender全然理解できてなかったです、ありがとうございます！

**Reply by yukiny0811** [reply]
なおりそうー？

**Reply by mm0o21** [reply]
行けましたー！ありがとうございます！
